File: hw/cañarte/u3/HW29Composite&TemplateMethod/Python/Composite/controller/Supervisor.py
import sys
import os
# Saray Cañarte, codecreators, DCCO - ESPE

sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))
from controller.Employee import Employee


class Supervisor(Employee):
    def __init__(self, name="not assigned yet"):
        super().__init__(name, "Supervisor")
        self.direct_reports = []

    def state_name(self):
        super().state_name()
        for report in self.direct_reports:
            report.state_name()

    def add(self, employee):
        self.direct_reports.append(employee)

    def remove(self, employee):
        self.direct_reports.remove(employee)

File: hw/CEDENOR/U3/HW29Composite&TemplateMethod/Java/CompositePattern/src/main/java/ec/edu/espe/compositepattern/controller/Supervisor.java
package ec.edu.espe.compositepattern.controller;
import java.util.Vector;
/**
 *
 * @author Nahomi Cedeño, codecreators, DCCO - ESPE
 */

abstract class Supervisor extends Employee {
    protected Vector directReports = new Vector();

    @Override
    public void stateName() {
        super.stateName(); 
        if (directReports.size() > 0) { 
            for (int i = 0; i < directReports.size(); ++i) {
                ((Employee) directReports.elementAt(i)).stateName();
            }
        }
    }

    public void add(Employee anEmployee) {
        this.directReports.addElement(anEmployee);
    }
}
